File: combinationsum.py
class Solution:
	# @param {integer[]} candidates
	# @param {integer} target
	# @return {integer[][]}
	def combinationSum3(self, k, target):
		def findcombo(start,curtarget,sol,remain):
			if curtarget ==0:
				if remain == 0:
					self.ret.append(sol[:])
				return
			if remain == 0:
				return
			for i in range(start,end):
				if i> start:
					if candidates[i] == candidates[i-1]:
						continue
				if candidates[i] <= curtarget:
					sol.append(candidates[i])
					findcombo(i+1,curtarget-candidates[i],sol,remain-1)
					sol.pop(-1)
		candidates = range(1,10)
		end = 9
		self.ret = []
		findcombo(0,target,[],k)
		return self.ret
s = Solution()
print s.combinationSum3(3,9)

File: combinationsum.cpp
#include "iostream"
#include "vector"
using namespace std;
class Solution {
public:
	vector<vector<int> > ret;
    vector<vector<int> > combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
     	vector<int> path;   
        helper(candidates,target,0,path);
        return ret;
    }
    void helper(vector<int> nums,int target,int start,vector<int>& path){
		if(target == 0){
			ret.push_back(path);
			return;
		}
    	for (int i = start; i < nums.size() && target>=nums[i]; ++i)
    	{
    		int num = nums[i];
    		path.push_back(num);
    		helper(nums,target-num,i,path);
    		path.pop_back();
    	}
    }
};
int main(int argc, char const *argv[])
{
	int myints[] = {7,6,3,2};
  	vector<int> nums (myints, myints + sizeof(myints) / sizeof(int) );
  	Solution s;
  	for(vector<int> v:s.combinationSum(nums,7)){
  		for(int num:v){
  			cout<<num<<endl;
  		}
  	}
	return 0;
}